In [ ]:
%autosave 0

Autosave disabled


In [ ]:
!wget https://github.com/alexeygrigorev/mlbookcamp-code/releases/download/chapter7-model/xception_v4_large_08_0.894.h5 -O clothing-model.h5

--2022-11-28 17:45:01--  https://github.com/alexeygrigorev/mlbookcamp-code/releases/download/chapter7-model/xception_v4_large_08_0.894.h5
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/256401220/0156a400-0049-11eb-8490-c0d01b48ea8c?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221128%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221128T224501Z&X-Amz-Expires=300&X-Amz-Signature=027ff4fd028d077f42c734c24b3d4228115ee9b978eda797da8a37e545b6161f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=256401220&response-content-disposition=attachment%3B%20filename%3Dxception_v4_large_08_0.894.h5&response-content-type=application%2Foctet-stream [following]
--2022-11-28 17:45:01--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/25640

Just want to be on at least 3.8:

In [ ]:
!python -V

Python 3.8.13


Want the latest 2.0 release:

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

tf.__version__

'2.8.2'

Loading model. Just saying there will not be GPU support and that it is not on the right CPU setup.

In [ ]:
model = keras.models.load_model("clothing-model.h5")

2022-11-28 17:46:27.654491: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:922] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node
Your kernel may have been built without NUMA support.
2022-11-28 17:46:27.746313: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda-11.4/lib64:
2022-11-28 17:46:27.746348: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1850] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2022-11-28 17:46:27.748453: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimi

Now we will predict on the pants image:

In [ ]:
!wget http://bit.ly/mlbookcamp-pants -O pants.jpg

--2022-11-28 17:47:57--  http://bit.ly/mlbookcamp-pants
Resolving bit.ly (bit.ly)... 67.199.248.10, 67.199.248.11
Connecting to bit.ly (bit.ly)|67.199.248.10|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://raw.githubusercontent.com/alexeygrigorev/clothing-dataset-small/master/test/pants/4aabd82c-82e1-4181-a84d-d0c6e550d26d.jpg [following]
--2022-11-28 17:47:58--  https://raw.githubusercontent.com/alexeygrigorev/clothing-dataset-small/master/test/pants/4aabd82c-82e1-4181-a84d-d0c6e550d26d.jpg
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23048 (23K) [image/jpeg]
Saving to: ‘pants.jpg’

pants.jpg           100%[===================>]  22.51K  --.-KB/s    in 0.005s  

2022-11-28 17:47:58 (4.84 MB/s) - ‘pants.jp

In [ ]:
from tensorflow.keras.preprocessing.image import load_img
from tensorflow.keras.applications.xception import preprocess_input

Loading the image into NumPy array and predicting:

In [18]:
img = load_img('pants.jpg', target_size=(299, 299))

x = np.array(img)
X = np.array([x])

X = preprocess_input(X)
X.shape

(1, 299, 299, 3)

In [23]:
preds = model.predict(X)
preds

array([[-1.8798645, -4.75631  , -2.3595324, -1.089265 ,  9.903785 ,
        -2.8261786, -3.6483104,  3.2411547, -2.6120958, -4.8520355]],
      dtype=float32)

In [24]:
classes = [
    'dress',
    'hat',
    'longsleeve',
    'outwear',
    'pants',
    'shirt',
    'shoes',
    'shorts',
    'skirt',
    't-shirt'
]

In [25]:
dict(zip(classes, preds[0]))

{'dress': -1.8798645,
 'hat': -4.75631,
 'longsleeve': -2.3595324,
 'outwear': -1.089265,
 'pants': 9.903785,
 'shirt': -2.8261786,
 'shoes': -3.6483104,
 'shorts': 3.2411547,
 'skirt': -2.6120958,
 't-shirt': -4.8520355}

## Convert Keras to TF-Lite

TF-Lite is large, so it includes TF-Lite. [Conversion overview](https://www.tensorflow.org/lite/models/convert)

In [27]:
# Convert the model
converter = tf.lite.TFLiteConverter.from_keras_model(model) # path to the saved model
tflite_model = converter.convert()

# Save the model.
with open('clothing-model.tflite', 'wb') as f:
  f.write(tflite_model)

2022-11-28 17:54:54.676427: W tensorflow/python/util/util.cc:368] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.


INFO:tensorflow:Assets written to: /tmp/tmpjyvhnr8r/assets


2022-11-28 17:55:07.205774: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:357] Ignored output_format.
2022-11-28 17:55:07.205823: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:360] Ignored drop_control_dependency.
2022-11-28 17:55:07.206407: I tensorflow/cc/saved_model/reader.cc:43] Reading SavedModel from: /tmp/tmpjyvhnr8r
2022-11-28 17:55:07.226052: I tensorflow/cc/saved_model/reader.cc:78] Reading meta graph with tags { serve }
2022-11-28 17:55:07.226082: I tensorflow/cc/saved_model/reader.cc:119] Reading SavedModel debug info (if present) from: /tmp/tmpjyvhnr8r
2022-11-28 17:55:07.309700: I tensorflow/cc/saved_model/loader.cc:228] Restoring SavedModel bundle.
2022-11-28 17:55:07.731360: I tensorflow/cc/saved_model/loader.cc:212] Running initialization op on SavedModel bundle at path: /tmp/tmpjyvhnr8r
2022-11-28 17:55:07.896232: I tensorflow/cc/saved_model/loader.cc:301] SavedModel load for tags { serve }; Status: success: OK. Took 689828 

Under the hood, Keras model is as Keras format `SavedModel` in `/tmp/tmpjyvhnr8r/assets`, then save as new format

In [30]:
!ls -lh | grep "clothing"

-rwxrwxrwx 1 andre andre  83M Dec  7  2021 clothing-model.h5
-rwxrwxrwx 1 andre andre  81M Nov 28 17:55 clothing-model.tflite


This is now a little more involved than loading to Keras:

In [31]:
# Using tensorflow package as it is simpler, will only use tensorflow-lite later
import tensorflow.lite as tflite

In [35]:
# Need interpreter to open model
interpreter = tflite.Interpreter(model_path='clothing-model.tflite')

# Now we need to load weights for model. Only Keras does this automatically
interpreter.allocate_tensors()

Need to figure out what inputs and outputs are. Normally Keras handled this:

`keras.Model(inputs, outputs)`

In [37]:
# Saved in:
interpreter.get_input_details()

[{'name': 'serving_default_input_8:0',
  'index': 0,
  'shape': array([  1, 299, 299,   3], dtype=int32),
  'shape_signature': array([ -1, 299, 299,   3], dtype=int32),
  'dtype': numpy.float32,
  'quantization': (0.0, 0),
  'quantization_parameters': {'scales': array([], dtype=float32),
   'zero_points': array([], dtype=int32),
   'quantized_dimension': 0},
  'sparsity_parameters': {}}]

`shape_signature` starting with -1 means we can feed as many images as we want at once.Image

What we really want though is `index`:

In [40]:
input_index = interpreter.get_input_details()[0]['index']

This is actually where the inputs should go. Now let's get the output details:

In [41]:
interpreter.get_output_details()

[{'name': 'StatefulPartitionedCall:0',
  'index': 229,
  'shape': array([ 1, 10], dtype=int32),
  'shape_signature': array([-1, 10], dtype=int32),
  'dtype': numpy.float32,
  'quantization': (0.0, 0),
  'quantization_parameters': {'scales': array([], dtype=float32),
   'zero_points': array([], dtype=int32),
   'quantized_dimension': 0},
  'sparsity_parameters': {}}]

We're looking at `'index': 229,`. Means we have 230 objects in our model and this is the last one

In [43]:
output_index = interpreter.get_output_details()[0]['index']

Using the model:

In [46]:
# Need to set input tensor in the input index
interpreter.set_tensor(input_index, X)
# Running inference
interpreter.invoke()
# Getting output
preds = interpreter.get_tensor(output_index)

Getting the semantic information same as before:

In [49]:
classes = [
    'dress',
    'hat',
    'longsleeve',
    'outwear',
    'pants',
    'shirt',
    'shoes',
    'shorts',
    'skirt',
    't-shirt'
]

dict(zip(classes, pred[0]))

{'dress': -1.8798645,
 'hat': -4.75631,
 'longsleeve': -2.3595324,
 'outwear': -1.089265,
 'pants': 9.903785,
 'shirt': -2.8261786,
 'shoes': -3.6483104,
 'shorts': 3.2411547,
 'skirt': -2.6120958,
 't-shirt': -4.8520355}

Alexey not sure why there isn't a simpler way to do this (or why it's not obvious)

There are some problems here, concenring *X*:
- Had to use `load_img()` from Tf
- Had preprocess using `preprocess_input()` in TF
    
Thus have to get these from somewhere else

## Removing TF Dependency

Let's see how Tensorflow [loads the image](https://www.tensorflow.org/api_docs/python/tf/keras/utils/load_img)
- Go to Source -> Can see Keras loads into PIL format
- Go to source of image loading `keras/keras/utils/image_utils.py` > `load_img()`

It's doing a lot in the source function, but it essentially loads with PIL

In [50]:
from PIL import Image

In [52]:
with Image.open('pants.jpg') as img:
    # Uses default interpolation method from Keras source
    img = img.resize((299, 299), Image.NEAREST)

/tmp/ipykernel_446/1303958241.py:3: DeprecationWarning: NEAREST is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.NEAREST or Dither.NONE instead.
  img = img.resize((299, 299), Image.NEAREST)


Let's see if that works:

In [53]:
x = np.array(img)
X = np.array([x])

X = preprocess_input(X)

In [54]:
interpreter.set_tensor(input_index, X)
interpreter.invoke()
preds = interpreter.get_tensor(output_index)

In [55]:
classes = [
    'dress',
    'hat',
    'longsleeve',
    'outwear',
    'pants',
    'shirt',
    'shoes',
    'shorts',
    'skirt',
    't-shirt'
]

dict(zip(classes, pred[0]))

{'dress': -1.8798645,
 'hat': -4.75631,
 'longsleeve': -2.3595324,
 'outwear': -1.089265,
 'pants': 9.903785,
 'shirt': -2.8261786,
 'shoes': -3.6483104,
 'shorts': 3.2411547,
 'skirt': -2.6120958,
 't-shirt': -4.8520355}

Now we have to fix `preprocess_input()`.
1. Go back to [repo](https://github.com/keras-team/keras)
2. Open the Xception file `keras/keras/applications/xception.py`
3. See `preprocess_input()`
    - `data_format` is `None` and the mode is `tf`
    - Loads with `keras/keras/applications/imagenet_utils.py` > `def preprocess_input(x, data_format=None, mode="caffe")`
4. Since mode is `tf`, go to `_preprocess_numpy_input()`
5. Use the code:

In [57]:
def preprocess_input(x):
    x /= 127.5
    x -= 1.0
    return x

In [ ]:
# Remember we still need floats
x = np.array(img, dtype='float32')
X = np.array([x])

X = preprocess_input(X)

## Simpler way of doing this

Go to Alexey's [Keras Image Helper](https://github.com/alexeygrigorev/keras-image-helper)
- Works for other models (not just xception)
    - xception
    - resnet50
    - vgg16
    - inception_v3

Using example code:

In [58]:
!pip install keras-image-helper

2754.37s - pydevd: Sending message related to process being replaced timed-out after 5 seconds


In [61]:
# Need to create_preprocesser with package
from keras_image_helper import create_preprocessor

preprocessor = create_preprocessor('xception', target_size=(299, 299))

# Can load with .from_url() or from_path() (local path) 
url = 'http://bit.ly/mlbookcamp-pants'
X = preprocessor.from_url(url)

Now we can predict just as before:

In [ ]:
x = np.array(img)
X = np.array([x])

X = preprocess_input(X)

In [62]:
interpreter.set_tensor(input_index, X)
interpreter.invoke()
preds = interpreter.get_tensor(output_index)

In [63]:
classes = [
    'dress',
    'hat',
    'longsleeve',
    'outwear',
    'pants',
    'shirt',
    'shoes',
    'shorts',
    'skirt',
    't-shirt'
]

dict(zip(classes, pred[0]))

{'dress': -1.8798645,
 'hat': -4.75631,
 'longsleeve': -2.3595324,
 'outwear': -1.089265,
 'pants': 9.903785,
 'shirt': -2.8261786,
 'shoes': -3.6483104,
 'shorts': 3.2411547,
 'skirt': -2.6120958,
 't-shirt': -4.8520355}

Taking a look inside keras-image-helper:
- Go to `keras-image-helper/keras_image_helper/preprocessors.py`
- Can see there are different preprocessors for the framework (TF, Caffe)
- Also has some classesfor applications:
    - `XceptionPreprocessor()`
    - `VGGPreprocessor()`
- Keeps this mapping from model to Preprocessor in `helper.py`
    - Only knows these 4 but it is easily extendable


## Running it all together

May want to restart notebook for this

In [1]:
import tensorflow.lite as tflite
from keras_image_helper import create_preprocessor

In [2]:
interpreter = tflite.Interpreter(model_path='clothing-model.tflite')
interpreter.allocate_tensors()

input_index = interpreter.get_input_details()[0]['index']
output_index = interpreter.get_output_details()[0]['index']

INFO: Created TensorFlow Lite XNNPACK delegate for CPU.


In [3]:
preprocessor = create_preprocessor('xception', target_size=(299, 299))

# Can load image with .from_url() or from_path() (local path) 
url = 'http://bit.ly/mlbookcamp-pants'
X = preprocessor.from_url(url)

In [4]:
interpreter.set_tensor(input_index, X)
interpreter.invoke()
preds = interpreter.get_tensor(output_index)

In [6]:
classes = [
    'dress',
    'hat',
    'longsleeve',
    'outwear',
    'pants',
    'shirt',
    'shoes',
    'shorts',
    'skirt',
    't-shirt'
]

dict(zip(classes, preds[0]))

{'dress': -1.8798653,
 'hat': -4.7563086,
 'longsleeve': -2.3595326,
 'outwear': -1.0892627,
 'pants': 9.903783,
 'shirt': -2.8261797,
 'shoes': -3.648309,
 'shorts': 3.2411575,
 'skirt': -2.6120954,
 't-shirt': -4.8520327}

## Last thing
We still use tensorflow.lite for tensorflow-lite. Can just use that package
- See [website](https://www.tensorflow.org/lite/guide) > Other Edge
    - Note: They no longer release pre-built tflite-runtime wheels for Windows and macOS. For these platforms, you should use the [full TensorFlow package](https://www.tensorflow.org/install/), or [build the tflite-runtime package from source](https://www.tensorflow.org/lite/guide/build_cmake_pip).

In [7]:
!python3 -m pip install tflite-runtime

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 3.5 MB/s eta 0:00:0000:0100:01


Now we change the imports. May want to restart here.

In [14]:
# Need to get interpreter specifically
import tflite_runtime.interpreter as tflite
from keras_image_helper import create_preprocessor

In [10]:
interpreter = tflite.Interpreter(model_path='clothing-model.tflite')
interpreter.allocate_tensors()

input_index = interpreter.get_input_details()[0]['index']
output_index = interpreter.get_output_details()[0]['index']

INFO: Created TensorFlow Lite XNNPACK delegate for CPU.


In [11]:
preprocessor = create_preprocessor('xception', target_size=(299, 299))

# Can load image with .from_url() or from_path() (local path) 
url = 'http://bit.ly/mlbookcamp-pants'
X = preprocessor.from_url(url)

In [12]:
interpreter.set_tensor(input_index, X)
interpreter.invoke()
preds = interpreter.get_tensor(output_index)

In [13]:
classes = [
    'dress',
    'hat',
    'longsleeve',
    'outwear',
    'pants',
    'shirt',
    'shoes',
    'shorts',
    'skirt',
    't-shirt'
]

dict(zip(classes, preds[0]))

{'dress': -1.8798653,
 'hat': -4.7563086,
 'longsleeve': -2.3595326,
 'outwear': -1.0892627,
 'pants': 9.903783,
 'shirt': -2.8261797,
 'shoes': -3.648309,
 'shorts': 3.2411575,
 'skirt': -2.6120954,
 't-shirt': -4.8520327}

Now we are ready to deploy to AWS lambda. Only need the small part of TF, TF-Lite, as it's own package.

Next:
- We will make this into a Python script